libraries/botbuilder-dialogs-adaptive/src/actions/forEach.ts

```diff
@@ -100,70 +104,88 @@ export class ForEach<O extends object = {}> extends Dialog<O> {
         return this;
     }
 
-    public getDependencies(): Dialog[] {
-        return this.actions;
-    }
-
-    public async beginDialog(sequence: SequenceContext, options: ForEachOptions): Promise<DialogTurnResult> {
-        // Ensure planning context
-        if (!(sequence instanceof SequenceContext)) { throw new Error(`${this.id}: should only be used within an AdaptiveDialog.`) }
+    public async beginDialog(dc: DialogContext, options: O): Promise<DialogTurnResult> {
+        // Evaluate list
         if (!this.list) { throw new Error(`${this.id}: no list expression specified.`) }
+        const list = this.list.evaluate(this.id, dc.state);
 
-        // Unpack options
-        let { list, offset } = options;
-        if (list == undefined) { list = this.list.evaluate(this.id, sequence.state) }
-        if (offset == undefined) { offset = 0 }
-
-        // Get next page of items
-        const item = this.getItem(list, offset);
-
-        // Update current plan
-        if (item !== undefined) {
-            sequence.state.setValue(this.valueProperty, item);
-            sequence.state.setValue(this.indexProperty, offset);
-            const changes: ActionChangeList = {
-                changeType: ActionChangeType.insertActions,
-                actions: []
-            };
-            this.actions.forEach((action) => changes.actions.push({ dialogStack: [], dialogId: action.id }));
-
-            // Add a call back into forEachPage() at the end of repeated actions.
-            // - A new offset is passed in which causes the next page of results to be returned.
-            changes.actions.push({
-                dialogStack: [],
-                dialogId: this.id,
-                options: {
-                    list: list,
-                    offset: offset + 1
+        // Count list items
+        let count: number = 0;
+        if (Array.isArray(list)) {
+            count = list.length;
+        } else if (typeof list == 'object') {
+            for (const key in list) {
+                if (list.hasOwnProperty(key)) {
+                    count++;
                 }
-            });
-            sequence.queueChanges(changes);
+            }
         }
 
-        return await sequence.endDialog();
+        // Initialize persisted state
+        dc.state.setValue(LIST_KEY, list);
+        dc.state.setValue(COUNT_KEY, count);
```

Comment:
we can get both of these values without promoting it into memory.  If you have a large list we are going to push a huge copy of the list just to iterate over it.  I changed this to just bind to the list on each item iteration.